src/tf_deadness_analysis.h

```diff
@@ -71,9 +87,40 @@ class DeadnessAnalysis {
   // instance of DeadnessAnalysis in `result`.
   static Status Run(const Graph& graph,
                     std::unique_ptr<DeadnessAnalysis>* result);
+
+  // For Data Flow ops, updates predicate_string
+  // Deadness is typically introduced by control flow ops. So, all the outgoing
+  // edges from the data flow op have the same deadness predicate ('And'
+  // Predicate of all its input predicates) and we can attach a predicate string
+  // to the data-flow node (predicate of its output edge). Control flow ops are
+  // assigned a placeholder predicate string (CONTROL_FLOW_PRED_STRING) .
+  virtual Status GetNodePredicate(const Node& node, string& pred_string) = 0;
+
+  static bool IsControlFlowPredString(string& predicate) {
+    return CONTROL_FLOW_PRED_STRING == predicate;
+  }
+
+  static bool IsTruePredString(string& predicate) {
+    return TRUE_PRED_STRING == predicate;
+  }
+
+  static void GetControlFlowPredString(string& predicate) {
+    predicate = CONTROL_FLOW_PRED_STRING;
```

Comment:
could be `const`

Reply:
minor: return not needed. the earlier functions do't have it, so for consistency.

Reply:
I mean `const string& predicate`

all the `could be const` comments mean the references could be const

Reply:
NVM, silly me